src/mono/mono/metadata/class-internals.h

```diff
@@ -405,7 +405,7 @@ struct _MonoMethodInflated {
 	} method;
 	MonoMethod *declaring;		/* the generic method definition. */
 	MonoGenericContext context;	/* The current instantiation */
-	MonoImageSet *owner; /* The image set that the inflated method belongs to. */
+	MonoGenericMemoryManager *owner; /* The mem manager that the inflated method belongs to. */
```

Comment:
I think this should be a `MonoMemoryManager *owner;` and you should downcast to `MonoGenericMemoryManager` in the memory manager utility functions.  Treat `MonoMemoryManager` like an abstract base class.

In the places where you need to do stuff with its ALCs, downcast in utility functions that worth with either a Singleton memory manager or a Generic one.

Reply:
Why is that better ? I think the best approach would be getting rid of these subclasses altogether, memory managers are not very numerous, so it doesn't save too much space, just complicates code and makes it more error prone. Subclasses are not very well supported by C.

Reply:
yea, ok.

@CoffeeFlux do we lose anything in collectability if we combine generic and singleton mem managers into one struct?

Reply:
So the mem manager functions take a MonoMemoryManager, but the caches are in the generic version, so no matter what the type of this field is, you need either casts or 2 variables.

Reply:
I think using a single memory manager struct is ok.

If we just have a single memory manager struct, do you think the caches should be in it?  I think you use memory managers in the JIT in some places - and I guess we will have to say that the caches in those have to be NULL.

Is it possible to confuse one memory manager for another and write something to a cache accidentally?

Maybe we at least need a kind enum on the mem manager struct that we can assert if it's a JIT mm the caches are null when we destroy the mm?

Reply:
The JIT memory manager is an extension of a normal memory manager, accessed through a pointer. So the same struct owns both metadata and JIT data.